hub/core/transform/transform.py

```diff
@@ -0,0 +1,148 @@
+            InvalidOutputDatasetError: If all the tensors of ds_out passed to transform don't have the same length. Using scheduler other than "threaded" with hub dataset having base storage as memory as ds_out will also raise this.
+            TensorMismatchError: If one or more of the outputs generated during transform contain different tensors than the ones present in 'ds_out' provided to transform.
+            UnsupportedSchedulerError: If the scheduler passed is not recognized. Supported values include: 'threaded' and 'processed'.
+        """
+        if isinstance(data_in, hub.core.dataset.Dataset):
+            data_in = get_dataset_with_zero_size_cache(data_in)
+
+        check_transform_data_in(data_in, scheduler)
+        check_transform_ds_out(ds_out, scheduler)
+
+        ds_out.flush()
+        initial_autoflush = ds_out.storage.autoflush
+        ds_out.storage.autoflush = False
+
+        tensors = list(ds_out.tensors)
+        compute_provider = get_compute_provider(scheduler, num_workers)
+
+        self.run_pipeline(data_in, ds_out, tensors, compute_provider, num_workers)
+        ds_out.storage.autoflush = initial_autoflush
+
+    def run_pipeline(
+        self,
+        data_in,
+        ds_out: hub.core.dataset.Dataset,
+        tensors: List[str],
+        compute: ComputeProvider,
+        num_workers: int,
+    ):
+        """Runs the pipeline on the input data to produce output samples and stores in the dataset."""
+        size = math.ceil(len(data_in) / num_workers)
+        slices = [data_in[i * size : (i + 1) * size] for i in range(num_workers)]
+
+        output_base_storage = get_base_storage(ds_out.storage)
+        metas_and_encoders = compute.map(
+            store_data_slice,
+            zip(slices, repeat(output_base_storage), repeat(tensors), repeat(self)),
+        )
+
+        all_tensor_metas, all_chunk_id_encoders = zip(*metas_and_encoders)
+        merge_all_tensor_metas(all_tensor_metas, ds_out)
+        merge_all_chunk_id_encoders(all_chunk_id_encoders, ds_out)
+
+
+def compose(transform_functions: List[TransformFunction]):
+    """Takes a list of transform functions and creates a pipeline out of them that can be evaluated using .eval"""
+    if not transform_functions:
+        raise TransformComposeEmptyListError
+    for fn in transform_functions:
+        if not isinstance(fn, TransformFunction):
+            raise TransformComposeIncompatibleFunction
+    return Pipeline(transform_functions)
+
+
+def compute(fn):
+    """Compute is a decorator for functions.
+    The functions should have atleast 2 argument, the first two will correspond to sample_in and samples_out.
+    There can be as many other arguments as required.
+    The output should be appended/extended to the second argument in a hub like syntax.
```

Comment:
instead of this wordy explanation, i recommend just defining an example function

```python

Example:
    def your_function(sample_in: Any, samples_out: Dataset, your_arg0, your_arg1=0):
        samples_out.your_tensor.append(your_arg0 * your_arg1)
```